FSE: Introduce a new block Posts Carousel

obenland opened the pull request:
#### Changes proposed in this Pull Request

* First pass at importing Newspack's Carousel block.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Make sure you have a post or two with a featured image
* Run `$ yarn && yarn run sync:newspack-blocks --nodemodules` to sync the blocks.
* Run `$ yarn dev` to build the plugin. 
* See if blog posts block and posts carousel block appear in the template selector.

Fixes #40979.

Patch: D42556-code

obenland committed:
Posts Carousel: Add block to import

obenland commented:
Upstream fix in https://github.com/Automattic/newspack-blocks/pull/437.

noahtallen commented on apps/full-site-editing/package.json:
```diff
@@ -81,6 +81,7 @@
 		"lodash": "*",
 		"moment": "*",
-		"newspack-blocks": "github:Automattic/newspack-blocks#v1.2.1"
+		"newspack-blocks": "github:Automattic/newspack-blocks#v1.2.1",
```
does the newspack version need updated? (not sure if this is just a hotfix :))

obenland replied:
No idea, but I updated it nonetheless :)

obenland committed:
Update newspack-blocks version

Props @noahtallen

noahtallen commented:
btw you might need to run `yarn` again to get the `yarn.lock` to update. Otherwise CI might not see the changes :)

But after that step is done, I think the step `yarn run sync:blog-posts-block` just needs to be `yarn` and then the plugin will be correctly installed and synced automatically!

obenland commented:
It should be noted that the block doesn't work for me in the front-end yet. but I suppose that's what #40981 is for.

obenland committed:
Update newspack-blocks version follow-up

retrofox commented on apps/full-site-editing/full-site-editing-plugin/blog-posts-block/editor.js:
```diff
@@ -10,16 +10,24 @@ import { addFilter } from '@wordpress/hooks';
  * Block name in the A8C\FSE context.
  */
-const blockName = 'a8c/blog-posts';
+const blogPostskName = 'a8c/blog-posts';
```
typo
```suggestion
const blogPostsName = 'a8c/blog-posts';
```

retrofox commented on apps/full-site-editing/full-site-editing-plugin/blog-posts-block/editor.js:
```diff
@@ -10,16 +10,24 @@ import { addFilter } from '@wordpress/hooks';
 function setBlockTransformationName( name ) {
-	return name !== 'newspack-blocks/homepage-articles' ? name : blockName;
+	if ( name === 'newspack-blocks/homepage-articles' ) {
+		return blogPostskName;
```
```suggestion
		return blogPostsName;
```

retrofox commented on apps/full-site-editing/full-site-editing-plugin/blog-posts-block/editor.js:
```diff
@@ -28,10 +36,15 @@ addFilter(
 
-registerBlockType( blockName, {
-	...settings,
+registerBlockType( blogPostskName, {
```
```suggestion
registerBlockType( blogPostsName, {
```

retrofox commented on apps/full-site-editing/full-site-editing-plugin/blog-posts-block/editor.js:
```diff
@@ -28,10 +36,15 @@ addFilter(
 	title: __( 'Blog Posts', 'full-site-editing' ),
 	category: 'layout',
 } );
+registerQueryStore( blogPostskName );
```
```suggestion
registerQueryStore( blogPostsName );
```

obenland committed:
Fix typo

Props @retrofox

retrofox commented:
> LGTM
> 
> ![image](https://user-images.githubusercontent.com/77539/80222253-f07ab000-861c-11ea-90cb-8eaa96507075.png)
> 
> Going to work on the newspack side for this.

[You already did it!](https://github.com/Automattic/newspack-blocks/pull/437)

obenland committed:
Load carousel view

retrofox reviewed (commented):
Let's remove the previews ✅ . This is our feature branch.

obenland committed:
FSE: Rename BPB to Newspack blocks (#41442)

* FSE: Rename BPB to Newspack blocks

* Update synced folder name

obenland committed:
Try using trunk to build plugin

obenland committed:
Remove Newspack version

marekhrabe commented on apps/full-site-editing/package.json:
```diff
@@ -81,6 +81,7 @@
 		"lodash": "*",
 		"moment": "*",
-		"newspack-blocks": "github:Automattic/newspack-blocks#v1.2.1"
+		"newspack-blocks": "github:Automattic/newspack-blocks",
```
removing the version from here has caused the CI to silently generate a faulty build where it has this definition with no version number:

```
// Autogenerated by apps/full-site-editing/bin/sync-newspack-blocks.sh.
define( 'NEWSPACK_BLOCKS__VERSION', '' );
// End autogenerated area.
```

obenland replied:
I removed the version in an attempt to make the CI integration work with the latest changes that are on master. I should have done a better job documenting that in the commit message.
And now that I look at it I realize I should have tried `master`.

marekhrabe replied:
`#master` might just work! It will probably still generate `define( 'NEWSPACK_BLOCKS__VERSION', 'master' );`

that's a bit better than being empty but I wonder if that's a good dev workflow. Let's do whatever gets us working block and then figure out how to update our processes and scripts, if needed.

obenland replied:
> Let's do whatever gets us working block

That's really the intention here. Before we can launch this block all upstream improvements will have to be shipped. Then we can switch this to that new version

obenland committed:
Use master for testing purposes

Will have to be updated with the new Newspack version that includes our upstream changes, before this can be merged

Also updates argument docs

retrofox committed:
FSE: sync-bash script agonistic from running-command. (#41491)

* fse: update sync bin script

* FSE: support --nodemodules MODE in sync bin script

* FSE: update `prepare` script with yarn command

retrofox committed:
FSE: build Newspacks block independently (#41553)

* fse: update sync bin script

* fse: add blog-posts-block sources files

* fse: add scripts to build blog-posts-block

* newspack-blocks: update blog_posts_block_args

* nespack-blocks: add carousel source files

* fse: add newspack carousel scripts

* newspack: pass custom args to BPB and carousel

* newspack-blocks: update docs

* newspack: fix --path mode in sync bin script

* Make `yarn dev` work

* Remove old files and update new ones

* Invert prefix logic

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland committed:
FSE: Make Posts Carousel block disable-able (#41568)

retrofox committed:
FSE: Update Newspack sync bash script cross different platforms. (#41603)

* fse: update sync bin script

* fse: handle sed command cross platforms

obenland committed:
Update newspack-blocks to v1.3.1

Includes all of our upstream changes

marekhrabe committed:
explain why you no longer see synced code in git

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/blog-posts-block-editor.js:
```diff
@@ -1,25 +1,29 @@
 /* eslint-disable import/no-extraneous-dependencies */
 /**
- * Internal dependencies
+ * External dependencies
```
we usually call these `WordPress dependencies`

retrofox replied:
```suggestion
 * WordPress dependencies
```

marekhrabe commented on apps/full-site-editing/package.json:
```diff
@@ -81,6 +85,7 @@
 		"moment": "*",
-		"newspack-blocks": "github:Automattic/newspack-blocks#v1.2.1"
+		"newspack-blocks": "github:Automattic/newspack-blocks#v1.3.1",
+		"swiper": "4.5.1"
```
I think this is extraneous. Since we have `newspack-blocks` as a dependency, all its dependencies will also be installed automatically and swiper is there.

marekhrabe replied:
#41669

marekhrabe commented:
I'm addressing some of my feedback in https://github.com/Automattic/wp-calypso/pull/41669

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/index.php:
```diff
@@ -0,0 +1,117 @@
+ * @param string $name Block name.
+ * @return array
+ */
+function blog_posts_block_args( $args, $name ) {
```
This function now handles both blocks, let's rename. Doing in #41669

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/index.php:
```diff
@@ -0,0 +1,117 @@
+	wp_register_style( $block_prefix . 'editor', $editor_style, array(), NEWSPACK_BLOCKS__VERSION );
+
+	// View script.
+	$script_data = require NEWSPACK_BLOCKS__BLOCKS_DIRECTORY . $block_prefix . 'view.asset.php';
+	wp_register_script(
+		$block_prefix . 'view',
+		plugins_url( NEWSPACK_BLOCKS__BLOCKS_DIRECTORY . $block_prefix . 'view.js', __FILE__ ),
+		$script_data['dependencies'],
+		$script_data['version'],
+		true
+	);
```
#41669

marekhrabe committed:
FSE: Posts Carousel: Small fixes (#41669)

* remove extraneous swiper dependency

* rename function to reflect the broadened scope

* don't enqueue view.js for carousel

obenland committed:
Remove some unnecessarily changed lines

obenland committed:
Bring back empty lines

obenland commented:
@marekhrabe @jeryj All dependencies for this PR are resolved, this should just need a final ✅ and merge.

marekhrabe reviewed (approved):
I have tested on dotcom simple and atomic, as well as pure FSE on clean WP install with no other plugins. Block has worked everywhere for me, with the exception of documented bugs in newspack-blocks. Desktop and mobile works too.

Our limitations for dotcom (automattician only) and atomic (completely disabled) were enforced well.

jeryj reviewed (approved):
Tested locally, on a simple site, and on Atomic with wpcomsh-v2.4.111 available. No console errors and everything works as expected! Some minor, known issues as @marekhrabe has pointed out.

- Locally Blog Posts and Article Carousel available 
- Blog Posts and Article Carousel available on sandboxed simple site
- Built and uploaded FSE for Atomic: no Article Carousel, but Blog Posts is available
